sky/clouds/cudo.py

```diff
@@ -0,0 +1,354 @@
+
+
+@clouds.CLOUD_REGISTRY.register
+class Cudo(clouds.Cloud):
+    """Cudo Compute"""
+    _REPR = 'Cudo'
+
+    _INDENT_PREFIX = '    '
+    _DEPENDENCY_HINT = (
+        'Cudo tools are not installed. Run the following commands:\n'
+        f'{_INDENT_PREFIX}  $ pip install cudo-compute')
+
+    _CREDENTIAL_HINT = (
+        'Install cudoctl and run the following commands:\n'
+        f'{_INDENT_PREFIX}  $ cudoctl init\n'
+        f'{_INDENT_PREFIX}For more info: '
+        # pylint: disable=line-too-long
+        'https://skypilot.readthedocs.io/en/latest/getting-started/installation.html'
+    )
+
+    _CLOUD_UNSUPPORTED_FEATURES = {
+        clouds.CloudImplementationFeatures.STOP: 'Stopping not supported.',
+        clouds.CloudImplementationFeatures.SPOT_INSTANCE: 'Spot is not supported, as runpod API does not implement spot .',
+        clouds.CloudImplementationFeatures.CUSTOM_DISK_TIER: 'Custom disk tier is currently not supported on Cudo Compute',
+        clouds.CloudImplementationFeatures.DOCKER_IMAGE:
+            ('Docker image is currently not supported on Cudo. '
+             'You can try running docker command inside the '
+             '`run` section in task.yaml.'),
+    }
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 60
+
+    PROVISIONER_VERSION = clouds.ProvisionerVersion.SKYPILOT
+    STATUS_VERSION = clouds.StatusVersion.SKYPILOT
+
+    _regions: List[clouds.Region] = []
+
+    @classmethod
+    def _unsupported_features_for_resources(
+        cls, resources: 'resources_lib.Resources'
+    ) -> Dict[clouds.CloudImplementationFeatures, str]:
+        """The features not supported based on the resources provided.
+
+        This method is used by check_features_are_supported() to check if the
+        cloud implementation supports all the requested features.
+
+        Returns:
+            A dict of {feature: reason} for the features not supported by the
+            cloud implementation.
+        """
+        del resources  # unused
+        return cls._CLOUD_UNSUPPORTED_FEATURES
+
+    @classmethod
+    def _max_cluster_name_length(cls) -> Optional[int]:
+        return cls._MAX_CLUSTER_NAME_LEN_LIMIT
+
+    @classmethod
+    def regions(cls) -> List[clouds.Region]:
+        if not cls._regions:
+            # Add the region from catalog entry
+            cls._regions = [
+                clouds.Region(...),
+            ]
+        return cls._regions
+
+    @classmethod
+    def regions_with_offering(cls, instance_type: Optional[str],
+                              accelerators: Optional[Dict[str, int]],
+                              use_spot: bool, region: Optional[str],
+                              zone: Optional[str]) -> List[clouds.Region]:
+        assert zone is None, 'Cudo does not support zones.'
+        del accelerators, zone  # unused
+        if use_spot:
+            return []
+        if instance_type is None:
+            # Fall back to default regions
+            regions = cls.regions()
+        else:
```

Comment:
This is not required, as `instance_type is not None` is guaranteed by the caller.